File: c++/PalindromePartition.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:    
    bool isPalindrome(int start, int end, string source) {
        while (end > start) {
            if (source[end] != source[start]) {
                return false;
            }
            --end;
            ++start;
        }
        return true;
    }
    
    void findPartition(int curIndex, string s, vector<string> &curList, vector<vector<string>> &output) {        
        if (curIndex == s.length()) {
            output.push_back(curList);
            return;
        }
        
        for (int i = curIndex; i < s.length(); ++i) {
            string subString = s.substr(curIndex, i - curIndex +1);
            if (isPalindrome(curIndex, i, s)) {
                curList.push_back(subString);
                findPartition(i+1, s, curList, output);
                curList.pop_back();
            }
        }
        return;
    }
    
    vector<vector<string>> partition(string s) {
        vector<vector<string>> output;
        vector<string> tempList;
        
        findPartition(0, s, tempList, output);
        
        return output;
    }
};

int main() {
	Solution sol;
	string input = "ababab";

	vector<vector<string>> output;

	output = sol.partition(input);

	cout << "[";
	for (auto list: output) {
		cout << "[";
		for (auto str : list) {
			cout << str << ", ";
		}
		cout << "], ";
	}
	cout << "]" << endl;

	return 0;
}

File: python3/PalindromePartition.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        
        self.output = []
        def isPalindrome(start, end):
            while (end > start):
                if (s[end] != s[start]):
                    return False
                end -= 1
                start += 1
            return True
            
        def findPartition(curIndex, curList):
            if (curIndex == len(s)):
                self.output.append(list(curList))
                return
            
            for i in range(curIndex, len(s)):
                subString = s[curIndex:i+1]
                if isPalindrome(curIndex, i):
                    curList.append(subString)
                    findPartition(curIndex + len(subString), curList)
                    curList.pop()
            return

        findPartition(0, [])
        
        return self.output
